### PR TITLE
Fix two skin-related crashes

### PR DESCRIPTION
**Pull request by garbear**
This PR contains two fixes:

* Fix segfault in guilib when `<itemlayout>` or `<focuslayout>` are missing or incomplete
* Fix segfault on skin reload due to accessing items after their owner has been destroyed

## How Has This Been Tested?
Fixes have been in my branch through most of the game OSD development.

## Types of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: xbmc/listproviders/StaticProvider.cpp [hunk 1]
```diff
@@ -117,6 +117,6 @@ bool CStaticListProvider::AlwaysFocusDefaultItem() const
 
 bool CStaticListProvider::OnClick(const CGUIListItemPtr &item)
 {
-  CGUIStaticItemPtr staticItem = std::static_pointer_cast<CGUIStaticItem>(item);
+  CGUIStaticItem *staticItem = static_cast<CGUIStaticItem*>(item.get());
```
Comment (tamland) [inline xbmc/listproviders/StaticProvider.cpp]: Something missing in this commit? This changes nothing

Comment (garbear) [reply xbmc/listproviders/StaticProvider.cpp]: The problem is `item`'s lifetime ends with the final call in this function. Acquiring a shared pointer means that item is used after `item`'s lifetime ends.

**Comment by notspiff** [comment]
Holding a shared ptr extends the item's lifetime. Now you make it dangling
instead which does not sound like an improvement.

10. des. 2017 17:54 skrev "Garrett Brown" <notifications@github.com>:

**Comment by garbear** [comment]
Dangling and no crash is better than dangling and crash

**Comment by tamland** [comment]
It's absolutely not. All this does is make it do something undefined instead the correct thing. A shared ptr doesn't dangling nor does any list control 'own' them (they are shared).

**Comment by garbear** [comment]
The behavior is only undefined if the object is used after being destroyed. Otherwise `delete this;` would be UB, yet our codebase has 15 instances of that.

**Comment by tamland** [comment]
Then what does casting a shared pointer, that's already in scope, to a raw pointer do?

**Review by xhaggi — APPROVED** [review]
Looks good

**Comment by garbear** [comment]
> Then what does casting a shared pointer, that's already in scope, to a raw pointer do?

Avoids a copy. The problem with a copy is that, due to RAII, the object's destructor is called when the ref count drops to zero. If we increase the ref count before the object should be destroyed, we call a destructor on a dangling object.

**Comment by tamland** [comment]
Then it's UB. If the share pointer is not in scope, you can not use it after the `item.get()`

> The behavior is only undefined if the object is used after being destroyed

Also not true. All un-synchronized access is undefined. Like deleting it while it's being used.

**Comment by garbear** [comment]
I'm confused. If that's true, then how is `delete this;` not UB?

**Comment by tamland** [comment]
It is UB, if it's not synchronized (and probably many other cases)

**Comment by garbear** [comment]
Gotchya. What about this? https://github.com/garbear/xbmc/commit/2ac65f8

I'm just eyeballing it here. I'll try applying soon this and see what happens.

**Comment by garbear** [comment]
No luck. I think I've solved a bug like this before. Got some history to look through...

**Comment by MartijnKaijser** [comment]
jenkins build and merge

**Comment by garbear** [comment]
Merging to fix https://trac.kodi.tv/ticket/17935 before release. This is an improvement regardless of how much refactoring should be done.

**Pull request merged**